state/prunning_service.go

```diff
@@ -0,0 +1,228 @@
+package state
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cometbft/cometbft/libs/service"
+)
+
+type PruningRequester int64
+
+const (
+	AppRequester                 PruningRequester = 0
+	DataCompanionRequester       PruningRequester = 1
+	ABCIResRetainHeightRequester PruningRequester = 2
+)
+
+var (
+	AppRetainHeightKey           = []byte("AppRetainHeightKey")
+	DataCompanionRetainHeightKey = []byte("DCRetainHeightKey")
+	ABCIResultsRetainHeightKey   = []byte("ABCIResRetainHeightKey")
+)
+
+type Pruner struct {
+	service.BaseService
+	logger log.Logger
+	// DB to which we save the retain heights
+	bs BlockStore
+	// State store to prune state from
+	stateStore    Store
+	pruningTicker *time.Ticker
```

Comment:
What's the benefit of using a ticker here as opposed to just configuring an interval and doing a `time.Sleep` between pruning operations? Would they have a similar effect?

I suppose as long as you don't create a hot loop of pruning operations with the ticker-based approach in cases where pruning takes longer than the pruning ticker timeout (i.e. > 10s) it should be okay.

Reply:
Hmm I did not think of that, that's  a good point. I don't think there is a way currently to prevent this behaviour. Epecially if we allow the operator to configure this parameter.  I initially did it with sleep but then chose this way because I found it cleaner to have a way to accept a signal to cancel the go routine and break the loop (which can be done even w/o a ticker).  But given your comment, it's much safer to just `time.Sleep`